polar-core/src/kb.rs

```diff
@@ -96,10 +96,13 @@ impl KnowledgeBase {
     }
 
     pub fn validate_rules(&self) -> Vec<Diagnostic> {
-        let mut diagnostics = self.validate_rule_calls();
+        let mut diagnostics = vec![];
         if let Err(e) = self.validate_rule_types() {
             diagnostics.push(Diagnostic::Error(e));
         }
+
+        diagnostics.append(&mut self.validate_rule_calls());
+
```

Comment:
why this change?

Reply:
because rule types are not dynamically generated validating rule calls before types with malformed resources/relations can produce an opaque "call to undefined rule `has_relation"`

Reply:
rather than special casing undefined rule calls for "built ins" re-ordering to validate types first will surface the missing required rule implementation error instead

Reply:
Great catch; would you mind adding comments to the code that capture that nuance? Otherwise, I am _definitely_ not going to remember that when I next come bulldozing through here 😆

Reply:
updated w/ comment reflecting the need to order rule type validation before call validation.